### PR TITLE
Unify `Dockerfile` naming and simplify + combine scripts

### PR DESCRIPTION
**Pull request by MarijnS95**

### CODE REVIEW

File: build-and-push-all.sh
Comment (rpavlik) [inline build-and-push-all.sh:15]: I think this image no longer exists.

**Comment by MarijnS95** [comment]
@rpavlik For this to work I should either modify the azure pipeline, or bump to GitHub Actions directly. What would you prefer? I'm personally a fan of doing one change at a time (for reviewability and quick PR throughput), and don't think updating Azure is too bad :)

File: build-and-push-all.sh
Comment (MarijnS95) [inline build-and-push-all.sh:16]: Speaking of fragile, this version got desynced with the Azure pipeline which was using `202201` here :)

**Comment by MarijnS95** [comment]
It looks like the Azure pipeline stopped triggering now that I modified the script (perhaps out of safety concerns), curious how it runs since I went for an array under `tags:` whereas code snippets suggest to oddly use a multiline codeblock instead.

File: build-one.sh
Comment (MarijnS95) [inline build-one.sh:29]: Two `if`s like this are a bit ugly, let's see if we can think of something better.

Comment (MarijnS95) [reply build-one.sh:29]: Should we perhaps always consume `$2`, and _force_ the caller to invoke eg:

```console
$ ./for-all.sh build
$ ./for-all.sh buildAndPush
```

Or something like that?

**Comment by oddhack** [comment]
Leaving aside the Rust image part, I don't understand what this PR is about - could you explain what the Azure stuff is doing, so I can get my head around this?

As far as the vulkan / asciidoctor-spec images are concerned, I'm unsure of the benefit of having dated / versioned images? The current setup in vulkan with a static container name whose contents are occasionally manually updated allows bumping the asciidoctor version globally across all active branches running CI at once, which is a net win IME though also has potential downsides.

**Comment by MarijnS95** [comment]
> Leaving aside the Rust image part, I don't understand what this PR is about - could you explain what the Azure stuff is doing, so I can get my head around this?

This PR specifically simplifies some scripts, makes the naming across dockerfiles consistent, and uniformly tags all images with both `-latest` and `-<date>` (hardcoded date) suffix.

I files this in a hurry as draft to show @rpavlik what I was working on, but didn't get to writing a nice overview even after "finishing" it and marking as ready for review. Is that something you'd like in the commit message and PR description before going further with this?

Azure is the CI, building the containers for testing and publishing them automatically on a push to `master`.

> As far as the vulkan / asciidoctor-spec images are concerned, I'm unsure of the benefit of having dated / versioned images? The current setup in vulkan with a static container name whose contents are occasionally manually updated allows bumping the asciidoctor version globally across all active branches running CI at once, which is a net win IME though also has potential downsides.

Same thought here, we can publish (and republish) those tags but never use them, just always stick to `-latest` (or would you want to omit that suffix entirely, because we're already using tags to identify the various Khronos-related docker images?). Or do you rather not have those tags at all?

**Comment by oddhack** [comment]
> > Leaving aside the Rust image part, I don't understand what this PR is about - could you explain what the Azure stuff is doing, so I can get my head around this?
> 
> This PR specifically simplifies some scripts, makes the naming across dockerfiles consistent, and uniformly tags all images with both `-latest` and `-<date>` (hardcoded date) suffix.
> 
> I files this in a hurry as draft to show @rpavlik what I was working on, but didn't get to writing a nice overview even after "finishing" it and marking as ready for review. Is that something you'd like in the commit message and PR description before going further with this?

Not needed, now that I understand the purpose. But I am not entirely certain that automatically updating the images on merge to the default branch is desirable behavior for my images. I've just been building locally and testing out of the local docker image cache before manually pushing to dockerhub, which should produce identical images - and yet I have a feeling of more control doing it this way. If the Dockerfiles were getting updated frequently automating the build/push process would make more sense, but it's been at most a couple of times/year for the Vulkan images.

> Same thought here, we can publish (and republish) those tags but never use them, just always stick to `-latest` (or would you want to omit that suffix entirely, because we're already using tags to identify the various Khronos-related docker images?). Or do you rather not have those tags at all?

Changing from 'asciidoctor-spec' to '...spec-latest' can get reflected in the vulkan repo default branches immediately, but it takes a while (possibly until just before merger to main) to reflect in the many development branches, so existing branches would tend to stick on the older non-latest image. So I'd like to continue with the current naming scheme for that and the vulkan-* images. @rpavlik has different considerations for the openxr-* images based on their needs AFAIK, so that's not a general statement.

I may also be a little confused on the difference between image 'name' and 'tag'. Having extra asciidoctor-spec-latest/asciidoctor-spec-date tags as aliases of the existing asciidoctor-spec doesn't affect our uses at all.

**Comment by MarijnS95** [comment]
> Not needed, now that I understand the purpose. But I am not entirely certain that automatically updating the images on merge to the default branch is desirable behavior for my images. 

Original behaviour is retained to only do this for OpenXR (other images aren't build-tested either). I can remove it from the Rust PR too - but we should probably at least build-test all the other images?

> Changing from 'asciidoctor-spec' to '...spec-latest' can get reflected in the vulkan repo default branches immediately, but it takes a while (possibly until just before merger to main) to reflect in the many development branches, so existing branches would tend to stick on the older non-latest image. So I'd like to continue with the current naming scheme for that and the vulkan-* images. @rpavlik has different considerations for the openxr-* images based on their needs AFAIK, so that's not a general statement.

It is an odd format as well, since theoretically all these image names should be part of the "repo" bit, and the tag is typically just a combination of versions that this was built on/from/for (usually with a bunch of tags aliasing the same image).

> I may also be a little confused on the difference between image 'name' and 'tag'. Having extra asciidoctor-spec-latest/asciidoctor-spec-date tags as aliases of the existing asciidoctor-spec doesn't affect our uses at all.

Yeah exactly, perhaps we should improve the scripts to not emit this for the non-OpenXR repos? Or perhaps they can be dropped across the board once OpenXR builds figure out how to automatically refresh the tags on every build?

Depends on how big of an issue breaking changes are going to be though, if it's not possible to affix oneself at a specific version and only bump it together with eventual fixups in CI scripts.

**Comment by rpavlik** [comment]
Yeah, some of this is more awkward because we have unrelated images in the "repo" part of the name, differentiated only by tag. I don't remember why I did this at first, possibly there was some account quota thing on docker hub?

**Comment by rpavlik** [comment]
> @rpavlik For this to work I should either modify the azure pipeline, or bump to GitHub Actions directly. What would you prefer? I'm personally a fan of doing one change at a time (for reviewability and quick PR throughput), and don't think updating Azure is too bad :)

re azure vs github actions: For most things I'd honestly prefer github actions now that they're in a usable state. They weren't when I started this work.

I also agree it's probably not a big deal to drop the "push" part from CI, it never worked right for me anyway and pushing from local is not that big of a deal and probably gives more control.

**Comment by MarijnS95** [comment]
> Yeah, some of this is more awkward because we have unrelated images in the "repo" part of the name, differentiated only by tag. I don't remember why I did this at first, possibly there was some account quota thing on docker hub?

Can you look that up @rpavlik? IMO it'd be cleaner if we could drop the `docker-images` bit from `khronosgroup/docker-images` (that's implied...) and make this `khronosgroup/asciidoctor-spec:latest` etc :)

> re azure vs github actions: For most things I'd honestly prefer github actions now that they're in a usable state. They weren't when I started this work.

Same thought, it'll be simpler and it can be tested on forks.

https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-Readme.md: Docker is installed so we don't even need to use the special docker actions in the CI script, but can call straight into the shell scripts created here. That way those are neatly reused and tested at the same time.

> I also agree it's probably not a big deal to drop the "push" part from CI, it never worked right for me anyway and pushing from local is not that big of a deal and probably gives more control.

Shall we do that in a separate followup PR though? Don't want to conflate this one :)

**Comment by MarijnS95** [comment]
On that note, friendly reminder to see if we can re-run / enable the changed Azure pipeline on this PR somehow, so that it's ready for merging.

**Comment by MarijnS95** [comment]
> I think it's OK in principle? As long as it isn't auto-pushing the asciidoctor-spec / vulkan-* images and there's still a way to invoke the manual equivalent of the old build-and-push-one.sh, I'm good.

I haven't changed any of this yet. In a future PR where we change out Azure for GH actions it should be good to exclusively build-test all the images, but leave _all_ publishing including `openxr` images to a manual action?

Please see https://github.com/KhronosGroup/DockerContainers/pull/15#discussion_r896129472 though, I changed the way this script is called. It'd be great if you can do a formal review of the code changes in this PR :)

> Regarding Azure I basically know nothing - years ago I had Azure CI configured on a Vulkan repo but have forgotten all the details, except that I had to go to a Microsoft website to see CI outputs. Probably the easiest way to re-run the Azure pipeline would be to force-push an empty commit?

I've pushed to this branch on multiple occasions, and have a hunch changing the scripts requires "maintainer approval" before they run? That doesn't explain why #13 ran, where the scripts have also been altered. Will try again.

**Comment by MarijnS95** [comment]
https://github.com/KhronosGroup/DockerContainers/pull/15#issuecomment-1154431341 And I wrote the answer myself :)

File: .azure-pipelines/build-template.yml
Comment (MarijnS95) [inline .azure-pipelines/build-template.yml:17]: Apparently this isn't going to work for the `buildAnPush`: https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/build/docker?view=azure-devops#why-does-docker-task-ignore-arguments-passed-to-buildandpush-command

Should we separate that into a `build`+`push` or drop `push` altogether as there's a preference for manual, even for OpenXR?

Comment (MarijnS95) [reply .azure-pipelines/build-template.yml]: Done :)

**Comment by MarijnS95** [comment]
CI is green now!

File: build-one.sh
Comment (oddhack) [inline build-one.sh:18]: If I'm reading this right, the script can either be invoked as

build-one.sh dockerfile-name

or

build-one.sh push dockerfile-name

? If so, would be good to update the comment at lines 5-7 above.

**Review by oddhack — CHANGES_REQUESTED** [review]
Other than the minor documentation issue with build-one.sh, I'm happy with this.

File: build-one.sh
Comment (MarijnS95) [reply build-one.sh:20]: Updated. Because of the `shift` it's the third argument.

**Review by oddhack — APPROVED** [review]

**Comment by oddhack** [comment]
Looks good to me. BTW can I ask why all the force-pushes? If commits accumulate they'll just get squashed out when we merge the MR.

**Comment by MarijnS95** [comment]
@oddhack that's just the workflow I'm used to, and I don't like the noise that ends up in the final squash-merged commit (titles of all the "fixup" commits).

I guess it makes it harder for you to see what changed in every push, correct?

File: README.md
Comment (rpavlik) [reply README.md:14]: Thanks

**Comment by oddhack** [comment]
> @oddhack that's just the workflow I'm used to, and I don't like the noise that ends up in the final squash-merged commit (titles of all the "fixup" commits).
> 
> I guess it makes it harder for you to see what changed in every push, correct?

Right - something changed, but it's harder to tell what. Not a big deal.

**Comment by rpavlik** [comment]
interesting, because GitLab explicitly has a feature to let you compare there across force pushes. I'm surprised GitHub doesn't.

Looks fine to me if it's fine to @oddhack . The OpenXR image is pretty stable so I rarely even think about this repo.

**Comment by MarijnS95** [comment]
Github does have a "Changed since last viewed" button but it disappears as soon as you've "viewed", on its own terms (i.e. reload the page and its gone). All you have then is the link under the word `force-pushed` (blue here):

image

But it's worthless if I ended up rebasing on latest `main` at the same time, for example. I see where this is coming from.

If we're all done here I guess we can merge? Get the Rust image in soon after that, and do a GitHub conversion separately.

We didn't end up discussing the "repo" vs "tag" part of the name, but IMHO that deserves its own distinct documented/annotated PR regardless (just like a migration to GH Actions).

**Pull request merged**
